### DISCUSSION TITLE
Reverse Proxy returning 502 on requests to app service using client certificate authentication

### DISCUSSION THREAD

**Discussion by sharvaridesai123** [dotnet/yarp#895] [Q&A] [answered]
We are using YARP to redirect traffic from a web application in Azure to an app service in Azure that uses MTLS authentication. The error received is 502 ( bad gateway).
 Code from the startup in  web application:
 **In ConfigureServices:**
 ```c#
    var routes = new[]
            {
                new ProxyRoute()
                {
                    RouteId = "route1",
                    ClusterId = "cluster1",
                    Match = new ProxyMatch
                    {
                        Path = "/api/{*any}"
                    }
                }
            };
            var serviceUrl= Configuration["serviceUrl"];
            var cert = GetClientCertificate();
            var clusters = new[]
            {
                new Cluster()
                {
                    Id = "cluster1",
                    SessionAffinity = new SessionAffinityOptions { Enabled = true, Mode = "Cookie" },
                    Destinations = new Dictionary<string, Destination>(StringComparer.OrdinalIgnoreCase)
                    {
                        { "destination1", new Destination() { Address = serviceurl} }
                    },
                    HttpClient = new ProxyHttpClientOptions { ClientCertificate = cert, MaxConnectionsPerServer = 10, SslProtocols = SslProtocols.Tls12 }
                }
            };

            services.AddReverseProxy()
                .LoadFromMemory(routes, clusters);
                
                 **In Configure:**
                 
                  app.UseEndpoints(endpoints =>
            {
               
                endpoints.MapReverseProxy();
            });

```

The destination service has Client certificate mode turned on to Required and the validation of the cert is done in the x-arr-clientcert header of the request.
What am I missing? The request that is routed to the service is a HttpGet.

**Comment by Tratcher** [comment, marked as answer]
First you'll want to capture the server logs to see what the full error was.
https://github.com/microsoft/reverse-proxy/blob/c3c0407b613286734cfb9d37102a933840a7f2d1/samples/ReverseProxy.Config.Sample/appsettings.json#L7-L13
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-5.0#configure-logging

**Reply by sharvaridesai123** [reply]
Turned on the debugging, here is what I get
HTTP Error 502.0 - Bad Gateway
The specified CGI application encountered an error and the server terminated the process.
**Most likely causes:**
The CGI application did not return a valid set of HTTP errors.
A server acting as a proxy or gateway was unable to process the request due to an error in a parent gateway.
 Looks like something is happening at the proxy level and the request is not reaching the destination

**Reply by sharvaridesai123** [reply]
The problem seems to be the way we are sending the certificate with YARP.

**Reply by Tratcher** [reply]
That's still the standard text for a 502 status, not the Asp.Net or yarp server logs.

**Reply by sharvaridesai123** [reply]
I have turned on the logging for YARP, where would the yarp log the errors? We have turned on AppInsights, did not see anything there either?

**Reply by Tratcher** [reply]
The default is console logs, but you can enable other outputs like ApplicationInsights as shown here:
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-5.0#built-in-logging-providers

**Reply by sharvaridesai123** [reply]
Got this from HttpProxy
An error occurred while sending the request. <--- The request was aborted. <--- The HTTP/2 server reset the stream. HTTP/2 error code 'HTTP_1_1_REQUIRED' (0xd).
 Changed the HttpVersion of the app service from 2.0 to 1.1 and it started working. Are the client certs not supported with HTTP/2?

**Reply by Tratcher** [reply]
HTTP/2 disallows lazy client certificate negotiation (e.g. per request), client certs have to be negotiated at the start of the connection. How do you have the client cert requirement configured on the destination?

**Reply by sharvaridesai123** [reply]
Client Cert Mode is Set to "Required" in the destination app service.

**Reply by Tratcher** [reply]
Hmm, that's something you'll need to investigate through app service.

**Reply by hchunter** [reply]
> HTTP/2 disallows lazy client certificate negotiation (e.g. per request), client certs have to be negotiated at the start of the connection. How do you have the client cert requirement configured on the destination?

What would cause the client cert to be lazy-loaded?  We retrieve the cert from the X509Store on the app service and then attach it using the ProxyHttpClientOptions in the cluster before it is loaded from memory.

```c#
using (var store = new X509Store(StoreName.My, StoreLocation.CurrentUser, OpenFlags.ReadOnly))
            {
                var certificate = store.Certificates
                    .OfType<X509Certificate2>()
                    .FirstOrDefault(c => c.Subject.Contains(certificateName, StringComparison.OrdinalIgnoreCase));
                if (certificate == null)
                {
                    throw new Exception($"Certificate {certificateName} not found!");
                }

                return certificate;
            }
```

**Reply by Tratcher** [reply]
Lazy client cert negotiation is a server setting, not a client setting. You'll have to investigate your configuration on the destination app.
